### DISCUSSION TITLE
Question about learning rate

### DISCUSSION THREAD

**Discussion by ethandeuxh** [mrdbourke/pytorch-deep-learning#1326] [Q&A]
Hello, so I'm following the "Learn PyTorch for deep learning in a day. Literally." video, and I have a question about this section : "49. Writing testing loop code".

The a and b parameters of the linear function could be a huge number. That's why I'm wondering if the learning rate could be adaptive depending maybe on the loss ?
Maybe I did understand the concept of learning rate wrongly, but, if for example I set the lr to 0.01, it would take AGES to reach a big number like one billion.

I did find this piece of documentation that may help this goal [ReduceLROnPlateau](https://docs.pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html) but I'm not sure it's the right way, I'm new to PyTorch after all.
Here is my code to showcase what I mean by adaptive learning rate (of course this is not working for obvious reason, one thing I am sure is that is not the right approach to do it)
```python
import random
import time
import torch
from torch import nn

# Fonction affine
a = random.random()*1_000_000_000
b = random.random()*1_000_000_000
print(a, b)

# Data
X = torch.arange(0, 1, 0.02).unsqueeze(dim=1) # 100 valeurs
Y = a*X + b # Fonction affine appliquée à X

# Séparation des données
# Entrainement (80%) / Vérification (20%)
train_data = int(0.8 * len(X))
X_train, Y_train = X[train_data:], Y[train_data:]
X_test, Y_test = X[:train_data], Y[:train_data]

#
class LinearRegressionModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Parameter(torch.tensor([0.0]))
        self.b = nn.Parameter(torch.tensor([0.0]))
    def forward(self, x:torch.Tensor) -> torch.Tensor:
        return self.a*x + self.b
model = LinearRegressionModule()

#
speed = 2
loss_function = nn.L1Loss()
optimizer = torch.optim.SGD(params=model.parameters(), lr=speed)

#
while True:
    model.train()
    loss = loss_function(model(X_train), Y_train)

    print(loss, speed)
    print(list(model.parameters()))
    print("------------------")

    if loss/2 > speed:
        speed *= 2
        optimizer = torch.optim.SGD(params=model.parameters(), lr=speed)
    elif loss < speed:
        speed /= 2
        optimizer = torch.optim.SGD(params=model.parameters(), lr=speed)
    else:
        model.eval()
        break

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    time.sleep(2)
model.eval()

print(list(model.parameters()))
```

Thank in advance for the help !

**Comment by ethandeuxh** [comment]
```bash
a=385864674.396986 b=755366088.6649951
model.a=890.7569 model.b=1000.6650
```
Look, right here I ran the program with 100 000 epochs and a learning rate of 0.01 which is common.
How are we supposed to hard code the value of the learning rate if we don't know how big the number we try to find is big.

```bash
a=972147724.4993763 b=274805074.7197851
model.a=963025600.0 model.b=281900032.0
```
Here I set the learning rate to 1 000 000, wich is better but not good at all, how are we supposed to set the learning rate according to a number we don't know.
Isn't there a way to adjust it, maybe according to the loss, this could greatly optimize the time complexity of the problem?

**Comment by ethandeuxh** [comment]
Oh, I came across the concept of standardization, put a matrix of large numbers around 0.
So I re-write the program to do so, it work quite well (with small or large numbers, doesn't matter).

With the standardization, a learning rate of 0.01 is acceptable, I just have to "de-standardize" at the end.
Here is the program :
```python
import random
import time
import torch
from torch import nn

# Fonction affine
a = random.random()*1_000_000_000_000
b = random.random()*1_000_000_000_000
print("--------------")
print(f"a = {a}")
print(f"b = {b}")

# Data
X = torch.arange(0, 1, 0.02).unsqueeze(dim=1)
Y = a*X + b

# Normalisation
X_mean = X.mean()
X_std = X.std()
Y_mean = Y.mean()
Y_std = Y.std()
X_norm = (X - X_mean) / X_std
Y_norm = (Y - Y_mean) / Y_std

# Séparation entrainement (80%) / vérification (20%)
train_data = int(0.8 * len(X))
X_train, Y_train = X_norm[:train_data], Y_norm[:train_data]
X_test, Y_test = X_norm[train_data:], Y_norm[train_data:]

# Modèle
class LinearRegressionModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Parameter(torch.tensor([0.0]))
        self.b = nn.Parameter(torch.tensor([0.0]))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.a*x + self.b

model = LinearRegressionModule()

# Options
loss_function = nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

# Boucle d'entrainement
start = time.time()
model.train()
for epoch in range(10_000):
    loss = loss_function(model(X_train), Y_train)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

end = time.time()

# Dé-normalisation
a_round = model.a.item() * (Y_std.item() / X_std.item())
b_round = Y_mean.item() + model.b.item() * Y_std.item() - a_round * X_mean.item()

print("--------------")
print(f"a = {a_round}")
print(f"b = {b_round}")
print("--------------")
print(f"Time = {end-start} seconds")
```

However I'm still interested about adaptive learning rate, if it's a thing or not, even if it's not common nor recommended practice.
Thank !

**Comment by eclctc** [comment]
Adaptive learning rate is most definitely a thing. I think it's more prevalent in things like reinforcement learning where you're agent doesn't necessarily know the space it's operating in. If you're really interested, you should check out the python library Optuna. It runs experiments on your model and gives you the ideal parameter set. A little bit of a learning curve, but super helpful in experimentation with your parameters :)

**Comment by akashcodes23** [comment]
Most common cause is learning rate — either too high (loss jumps around) or too low (barely moves). Try:

```
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)  # start here
```
Also check: (1) Are you calling optimizer.zero_grad() before loss.backward()? (2) Is your data normalized? Raw pixel values (0–255) without normalization will stall training. (3) Try lr=0.01 and see if loss moves — then tune down.
